ostd/src/sync/rcu/mod.rs

```diff
@@ -2,105 +2,307 @@
-        };
-        RcuReclaimer { ptr: old_ptr }
+impl<P: OwnerPtr + Send, const MAYBE_UNINIT: bool> Rcu_<P, MAYBE_UNINIT> {
+    /// Replaces the current pointer with a new pointer.
+    ///
+    /// This function updates the pointer to the new pointer regardless of the
+    /// original pointer. If the original pointer is not NULL, it will be
+    /// dropped after the grace period.
+    pub fn update(&self, new_ptr: P) {
+        let new_ptr = <P as OwnerPtr>::into_raw(new_ptr).cast_mut();
+        let old_raw_ptr = self.ptr.swap(new_ptr, AcqRel);
+
+        if let Some(p) = NonNull::new(old_raw_ptr) {
+            // SAFETY: It was previously returned by `into_raw`.
+            unsafe { delay_drop::<P>(p) };
+        }
     }
 }
 
-#[clippy::has_significant_drop]
-#[must_use]
-pub struct RcuReadGuard<'a, P: OwnerPtr> {
-    obj: &'a <P as OwnerPtr>::Target,
-    rcu: &'a Rcu<P>,
+/// A guard that allows read-only access to the initialized data protected
+/// by the RCU mechanism.
+pub struct RcuReadGuard<'a, P: OwnerPtr, const MAYBE_UNINIT: bool> {
+    /// If maybe uninitialized, the pointer can be NULL.
+    obj_ptr: *mut <P as OwnerPtr>::Target,
+    rcu: &'a Rcu_<P, MAYBE_UNINIT>,
+    _inner_guard: DisabledPreemptGuard,
 }
 
-impl<'a, P: OwnerPtr> Deref for RcuReadGuard<'a, P> {
+// Initialized RCU guard can be dereferenced.
+impl<P: OwnerPtr> Deref for RcuReadGuard<'_, P, false> {
     type Target = <P as OwnerPtr>::Target;
 
     fn deref(&self) -> &Self::Target {
-        self.obj
+        // SAFETY: Since the preemption is disabled, the pointer is valid
+        // because other writers won't update the pointer until this task
+        // passes the quiescent state.
+        // And this pointer is not NULL.
+        unsafe { &*self.obj_ptr }
     }
 }
 
-#[repr(transparent)]
-pub struct RcuReclaimer<P> {
-    ptr: P,
+// Maybe uninitialized RCU guard can be dereferenced after checking.
+impl<'a, P: OwnerPtr> RcuReadGuard<'a, P, true> {
+    /// Get the initialized read guard.
+    ///
+    /// If the RCU object is not initialized, this function will return
+    /// [`Err`] with itself. Otherwise a dereferenceable read guard will be
+    /// returned.
+    pub fn check(self) -> Result<RcuReadGuard<'a, P, false>, Self> {
+        if self.obj_ptr.is_null() {
+            return Err(self);
+        }
+        Ok(RcuReadGuard {
+            obj_ptr: self.obj_ptr,
+            // SAFETY: It is initialized. The layout is the same.
+            rcu: unsafe { core::mem::transmute::<&Rcu_<P, true>, &Rcu_<P, false>>(self.rcu) },
+            _inner_guard: self._inner_guard,
+        })
+    }
 }
 
-impl<P: Send + 'static> RcuReclaimer<P> {
-    pub fn delay(mut self) {
-        let ptr: P = unsafe {
-            let ptr = core::mem::replace(&mut self.ptr, core::mem::uninitialized());
+impl<P: OwnerPtr + Send, const MAYBE_UNINIT: bool> RcuReadGuard<'_, P, MAYBE_UNINIT> {
+    /// Try to replaces the already read pointer with a new pointer.
```

Comment:
[`Try` -> `Tries`](https://rust-lang.github.io/rfcs/0505-api-comment-conventions.html#formatting):
> The summary line should be written in third person singular present indicative form. Basically, this means write “Returns” instead of “Return”.